tests/spiped_nc/graceful_shutdown.c

```diff
@@ -0,0 +1,96 @@
+#include <signal.h>
+#include <stdlib.h>
+
+#include "events.h"
+#include "warnp.h"
+
+#include "graceful_shutdown.h"
+
+struct graceful_shutdown_cookie {
+	void (* begin_shutdown)(void *);
```

Comment:
Also, we usually have callbacks return `int`, so that they can report if something goes fatally wrong.